File: components/themetoggle/themetoggle.js
// Theme (light / dark / system) switcher. Pairs with themetoggle.templ and the
// blocking InitScript in <head>; this file owns changes after load.
(function () {
  const KEY = "theme";
  const media = window.matchMedia("(prefers-color-scheme: dark)");

  function stored() {
    try {
      return localStorage.getItem(KEY) || "system";
    } catch (e) {
      return "system";
    }
  }

  function apply(theme) {
    const dark = theme === "dark" || (theme === "system" && media.matches);
    const root = document.documentElement;
    root.classList.toggle("dark", dark);
    root.style.colorScheme = dark ? "dark" : "light";
    root.dataset.theme = theme;
    syncGroups(theme);
  }

  // Server-rendered menus default to "system"; mirror the stored choice on
  // every group so the checkmark is right on load and after htmx swaps.
  function syncGroups(theme) {
    document.querySelectorAll("[data-tui-themetoggle]").forEach((group) => {
      group.setAttribute("data-tui-dropdownmenu-radio-value", theme);
      group.querySelectorAll("[data-tui-dropdownmenu-radio-item]").forEach((item) => {
        const checked = item.getAttribute("data-tui-dropdownmenu-radio-value") === theme;
        item.toggleAttribute("data-checked", checked);
        item.toggleAttribute("data-unchecked", !checked);
        item.setAttribute("aria-checked", checked ? "true" : "false");
      });
    });
  }

  function set(theme) {
    try {
      localStorage.setItem(KEY, theme);
    } catch (e) {}
    apply(theme);
  }

  document.addEventListener("dropdownmenu-value-change", (e) => {
    const group = e.target.closest && e.target.closest("[data-tui-themetoggle]");
    if (!group || !e.detail || !e.detail.value) return;
    set(e.detail.value);
  });

  // Follow the OS while in system mode.
  media.addEventListener("change", () => {
    if (stored() === "system") apply("system");
  });

  // Other tabs changing the preference.
  window.addEventListener("storage", (e) => {
    if (e.key === KEY) apply(stored());
  });

  // Sync freshly swapped-in menus right before they open.
  document.addEventListener("dropdownmenu-open-change", (e) => {
    if (e.detail && e.detail.open && e.target.querySelector("[data-tui-themetoggle]")) {
      syncGroups(stored());
    }
  });

  function init() {
    apply(stored());
  }
  if (document.readyState === "loading") {
    document.addEventListener("DOMContentLoaded", init);
  } else {
    init();
  }
  document.addEventListener("htmx:afterSettle", () => syncGroups(stored()));
})();

File: components/themetoggle/themetoggle.go
package themetoggle

import (
	"context"
	"io"

	"github.com/a-h/templ"

	"templ-app/components/dropdownmenu"
)

// triggerAttrs merges the dropdown trigger attributes with caller extras and
// the relative positioning the stacked sun/moon icons need.
func triggerAttrs(ctx context.Context, extra templ.Attributes) templ.Attributes {
	attrs := templ.Attributes{"aria-label": "Toggle theme", "data-tui-themetoggle-trigger": true}
	for k, v := range extra {
		attrs[k] = v
	}
	for k, v := range dropdownmenu.Trigger(ctx) {
		attrs[k] = v
	}
	return attrs
}

// initJS runs synchronously in <head> before the body is parsed, so a stored
// dark preference never flashes the light theme. It mirrors the resolve logic
// in themetoggle.js; keep the two in sync.
const initJS = `(function(){try{var t=localStorage.getItem("theme");var d=t==="dark"||((!t||t==="system")&&window.matchMedia("(prefers-color-scheme: dark)").matches);var r=document.documentElement;r.classList.toggle("dark",d);r.style.colorScheme=d?"dark":"light";}catch(e){}})();`

// InitScript renders the blocking theme bootstrap. Place it in <head> right
// after the stylesheet link.
func InitScript() templ.Component {
	return templ.ComponentFunc(func(ctx context.Context, w io.Writer) error {
		nonce := templ.GetNonce(ctx)
		open := "<script"
		if nonce != "" {
			open += ` nonce="` + templ.EscapeString(nonce) + `"`
		}
		_, err := io.WriteString(w, open+">"+initJS+"</script>")
		return err
	})
}
